kmir/src/kmir/kdist/mir-semantics/rt/data.md

```diff
@@ -1289,9 +1325,42 @@ rule <k> rvalueNullaryOp(nullOpAlignOf, TY)
 
 #### Other operations
 
-`binOpOffset`
+The unary operation `unOpPtrMetadata`, when given a reference to an array or slice, will return the array length of the slice length (which is dynamic, not statically known), as a `usize`.
+
+```k
+  rule <k> #applyUnOp(unOpPtrMetadata, typedValue(Reference(OFFSET, place(local(I), PROJECTIONS), _), _, _)) => #arrayLength({LOCALS[I]}:>TypedValue, PROJECTIONS)
+        ...
+       </k>
+       <locals> LOCALS </locals>
+    requires OFFSET ==Int 0
+     andBool 0 <=Int I
+     andBool I <Int size(LOCALS)
+     andBool isTypedValue(LOCALS[I])
+    [preserves-definedness] // LOCALS indexing checked
+
+  rule <k> #applyUnOp(unOpPtrMetadata, typedValue(Reference(OFFSET, place(LOCAL, PROJECTIONS), _), _, _)) => #arrayLength(#localFromFrame({STACK[OFFSET -Int 1]}:>StackFrame, LOCAL, OFFSET), PROJECTIONS)
+        ...
+       </k>
+       <stack> STACK </stack>
+    requires 0 <Int OFFSET
+     andBool OFFSET <=Int size(STACK)
+     andBool isStackFrame(STACK[OFFSET -Int 1])
+
+  syntax Evaluation ::= #arrayLength( TypedValue , ProjectionElems )
+
+  rule <k> #arrayLength(typedValue(Range(LIST), _, _), .ProjectionElems) => typedValue(Integer(size(LIST), 64, false), TyUnknown, mutabilityNot) ... </k>
+
+  syntax KItem ::= #arrayLength()
 
-`unOpPtrMetadata`
+  rule <k> #arrayLength(TV, PROJS) => #readProjection(TV, PROJS) ~> #arrayLength() ... </k>
+    requires notBool PROJS ==K .ProjectionElems
```

Comment:
A word of warning: I had this as an `[owise]` rule first, and got loads of branching on the `thunk` default rule for `Evaluation`s. We cannot use `owise` on rules over `Evaluation`.